# Newton法

## $\sqrt{2}$に収束する漸化式

$$
a_{n + 1} = \frac{1}{2}\left( a_n + \frac{2}{a_n} \right),\quad a_0 > 0
$$

In [1]:
a = 1.0
print(a)
for n in range(10):
    a = 0.5 * (a + 2/a)
    print(a)

1.0
1.5
1.4166666666666665
1.4142156862745097
1.4142135623746899
1.414213562373095
1.414213562373095
1.414213562373095
1.414213562373095
1.414213562373095
1.414213562373095


## Newton法

関数$f(x)$に対して$f(x) = 0$の解を求めたい．漸化式
$$
x_{n+1} = x_n - \frac{f(x_n)}{f'(x_n)}
$$
によって数列を計算する．

In [2]:
def f(x):
    return x**2 - 2

def df(x):
    return 2*x

初期値と反復回数を設定する

In [3]:
x = 2.0
N = 10

ニュートン法で数列を生成する

In [4]:
print(x)
for n in range(N):
    x = x - f(x) / df(x)
    print(x)

2.0
1.5
1.4166666666666667
1.4142156862745099
1.4142135623746899
1.4142135623730951
1.414213562373095
1.4142135623730951
1.414213562373095
1.4142135623730951
1.414213562373095


初期値を$x = 2$とした場合，4ステップ目で既に$1.41421356$まで一致している．5ステップ目以降は二つの数値が入れ替わりに
十分に収束しているようである

精度の改善が見られないのに反復を繰り返すのは効率が悪い．反復終了の判定条件をつけよう．
ここでは残差の絶対値が一定値以下になったら反復を停止して近似解を出力，
または一定回数の反復で終了条件を満たさなかったらエラーを返すことにする．
判定条件にはxの増分や相対残差，またはそれらの組み合わせを用いることもできる．

In [5]:
def Newton(x, f, df, eps=1.0e-8, maxiter=10):
    y = x
    for n in range(maxiter):
        y = y - f(y) / df(y)
        if (abs(f(y)) < eps):
            return y
    print("収束しなかった")

In [6]:
Newton(2, f, df, eps=1e-14)

1.4142135623730951

計算精度には限界がある．倍精度で残差を$10^{-16}$まで小さくするのは望めない．

In [7]:
Newton(2, f, df, eps=1e-16)

収束しなかった


## パッケージを使う

詳しくは
https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.newton.html#scipy.optimize.newton

In [8]:
from scipy.optimize import newton
newton(f, 2)

1.4142135623730954

## 複素ニュートン法

ニュートン法は複素関数にも拡張できる．例として$1$の$3$乗根を求めてみる．

In [9]:
def f(z):
    return z**3 - 1

def df(z):
    return 3*z**2

In [10]:
Newton(-1+1.j, f, df)

(-0.4999999999999555+0.8660254037846933j)